### PR TITLE
avoid explicitly casting low precision inputs to fp32 in norm

### PR DESCRIPTION
**Pull request by ngimel**
Per title. Now `norm` with fp16/bfloat16 inputs and fp32 outputs on cuda won't do explicit cast

### CODE REVIEW

File: test/test_linalg.py [hunk 1]
```diff
@@ -1562,6 +1562,24 @@ def check(op):
...
                     run_test_case(input, ord, dim, keepdim)
 
+
+    @onlyCUDA
+    @dtypes(torch.bfloat16, torch.float16)
+    def test_norm_fused_type_promotion(self, device, dtype):
+        x = torch.randn(10, device=device, dtype=dtype)
+
+        def profile_and_check(fn, x, kwargs, fn_name):
+            with torch.profiler.profile(activities=(torch.profiler.ProfilerActivity.CPU,)) as p:
```
Comment (zasdfgbnm) [inline test/test_linalg.py:1572]: `torch.profiler` is amazing

File: aten/src/ATen/native/ReduceOps.cpp [hunk 2]
```diff
@@ -1030,7 +1030,10 @@ static Tensor& norm_out(Tensor &result, const Tensor &self, const optional<Scala
 
   ScalarType out_dtype = result.defined() ? result.scalar_type() : (opt_dtype.has_value() ? opt_dtype.value() : toValueType(self.scalar_type()));
 
-  auto iter = make_reduction("norm", result, self, dim, keepdim, in_dtype, out_dtype);
+// omit in_dtype in the following call, to avoid make_reduction explicitly casting input to out_dtype
+  auto iter = isComplexType(self.scalar_type()) ?
+      make_reduction("vector_norm", result, self, dim, keepdim, in_dtype, out_dtype) :
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/ReduceOps.cpp:1035]: Why the name for complex is different? `vector_norm` vs `norm`

**Review by zasdfgbnm — APPROVED** [review]

File: aten/src/ATen/native/ReduceOps.cpp (hunk 2 above)
Comment (ngimel) [reply aten/src/ATen/native/ReduceOps.cpp:1035]: No reason, just copy paste. I'll fix it, thanks for flagging.

Comment (ngimel) [inline aten/src/ATen/native/ReduceOps.cpp:1035]:
```suggestion
      make_reduction("norm", result, self, dim, keepdim, in_dtype, out_dtype) :
```

**Pull request closed without merging**